### PR TITLE
Update widgets when task starts or becomes due

### PR DESCRIPTION
**Pull request by Trogel**
This is an attempt to fix issue #17.

I hope it is not too bad style to talk to the widgets from the notification-related `ActionService`.

### CODE REVIEW

File: opentasks/src/main/java/org/dmfs/tasks/homescreen/TaskListWidgetUpdaterService.java [hunk 1]
```diff
@@ -233,7 +233,7 @@ public RemoteViews getViewAt(int position)
                 row.setTextViewText(android.R.id.text1, mDueDateFormatter.format(dueDate, DateFormatContext.WIDGET_VIEW));
 
                 // highlight overdue dates & times
-                if ((!dueDate.allDay && dueDate.before(mNow) || dueDate.allDay
+                if ((!dueDate.allDay && Time.compare(dueDate, mNow) <= 0 || dueDate.allDay
```
Comment (dmfs) [inline opentasks/src/main/java/org/dmfs/tasks/homescreen/TaskListWidgetUpdaterService.java]: This looks unrelated to the issue. Why did you make this change?

File: opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+
+/**
+ * A {@link TaskAction} that updates the widgets.
+ *
+ * @author Trogel
+ */
+public final class UpdateWidgetsAction implements TaskAction
+{
+    private final static String TAG = "UpdateWidgetsAction";
```
Comment (dmfs) [inline opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java]: Please remove unused field.

File: opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    public UpdateWidgetsAction()
+    {
+    }
+
+
+    @Override
+    public void execute(Context context, ContentProviderClient contentProviderClient, RowDataSnapshot<TaskContract.Instances> rowSnapshot, Uri taskUri)
+    {
+        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
+        final ComponentName componentName = new ComponentName(context, TaskListWidgetProvider.class);
```
Comment (dmfs) [inline opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java]: Currently we have two widget providers (mostly for backwards compatibility). Please ensure to update both.

**Comment by dmfs** [comment]
Thanks for the PR. Nice solution. I don't have any issues with using the notification actions. This mechanism is sort of proof of concept for a more generic plug-in mechanism. So it's nice to see we can use it for other use cases.

File: opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java (hunk 2 above)
Comment (Trogel) [reply opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java]: Oh, that was a relict of some debug logging. Removed.

File: opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java (hunk 3 above)
Comment (Trogel) [reply opentasks/src/main/java/org/dmfs/tasks/actions/UpdateWidgetsAction.java]: Sorry, did not see the other. Hope I got it right.

File: opentasks/src/main/java/org/dmfs/tasks/homescreen/TaskListWidgetUpdaterService.java (hunk 1 above)
Comment (Trogel) [reply opentasks/src/main/java/org/dmfs/tasks/homescreen/TaskListWidgetUpdaterService.java]: While working on this fix I observed some cases where the update was processed right within the second of the due date. I noticed that in those cases the task did not turn red but remained white in the widget. That seemed wrong. The reason was that dueDate was not before but exactly equal to mNow – at least in `android.text.format.Time`'s second precision.

Therefore I changed this comparison essentially from `dueDate < mNow` to `dueDate <= mNow`. Sorry, if the commit message was not clear about my intention.

**Pull request merged**